pkg/util/provider/machinecontroller/machine_safety.go

```diff
@@ -264,16 +265,16 @@ func (c *controller) checkMachineClass(ctx context.Context, machineClass *v1alph
 	for machineID, machineName := range listMachineResponse.MachineList {
 		machine, err := c.machineLister.Machines(c.namespace).Get(machineName)
 
-		if err != nil && !apierrors.IsNotFound(err) {
-			// Any other types of errors
-			klog.Errorf("SafetyController: Error while trying to GET machines. Error: %s", err)
-		} else if err != nil || machine.Spec.ProviderID != machineID {
-
-			// If machine exists and machine object is still been processed by the machine controller
-			if err == nil &&
-				(machine.Status.CurrentStatus.Phase == "" || machine.Status.CurrentStatus.Phase == v1alpha1.MachineCrashLoopBackOff) {
-				klog.V(3).Infof("SafetyController: Machine object %q with backing nodeName %q , providerID %q is being processed by machine controller, hence skipping", machine.Name, getNodeName(machine), getProviderID(machine))
-				continue
+		if apierrors.IsNotFound(err) || err == nil {
+			if err == nil {
+				if machine.Spec.ProviderID == machineID {
+					continue
+				}
+				// machine obj is still being processed by the machine controller
+				if machine.Status.CurrentStatus.Phase == "" || machine.Status.CurrentStatus.Phase == v1alpha1.MachineCrashLoopBackOff {
+					klog.V(3).Infof("SafetyController: Machine object %q with backing nodeName %q , providerID %q is being processed by machine controller, hence skipping", machine.Name, getNodeName(machine), getProviderID(machine))
+					continue
+				}
```

Comment:
The first `if apierrors.IsNotFound(err) || err == nil` nests an `if err == nil` which in turn nests two if statements which will do something and continues the loop skipping everything below it. So, why not make that check of `err == nil` independent before the first if block; followed by an if block to check the NotFound error?

Reply:
are you suggesting this

```
if (no error){
 // checks for continue
}
if (NotFound error){
// dummy object creation and delete
}else{
//any other error
}
```

In this flow we'll miss the case where `err==nil && machine.Spec.ProviderID!=machineID && (machine not reconciled by MC)` , and we want to remove the VM for this case also.
Yes to solve that we could follow below flow which is the same flow as I have pushed , with just the if block taken out as you suggested.

```
if (no error){
 // checks for continue
}
if (NotFound error) or (no error){
// dummy object creation and delete
}else{
//any other error
}
```

Let me know your preference :)